### PR TITLE
generate input objects, start using Nullable

### PR DESCRIPTION
**Pull request by davidbloss**
Resolves [#573](https://github.com/OpsLevel/team-platform/issues/573)

### 🚨 This PR contains BREAKING CHANGES for the upcoming `v2025` opslevel-go release 🚨 

Run `task gen` to see new code generated in `input.go`

### Noteworthy Changes

- ❗The `*Nullable[<type>]` type now wraps optional fields of most structs in `input.go`. This sweeping change enables setting input object fields to JSON `null`, resolves ambiguity during API update calls, and completes a planned major refactor before the next major release of opslevel-go.
- In `scalar.go`, a new `UnmarshalJSON()` for `Nullable` types was added to correctly unmarshal JSON values to this custom type.
- `task gen` now uses the template in `templates/inputObjects.tpl` to generate everything in `input.go`
- In `gen.go`: `enumExamples`, `listExamples`, and `scalarExamples` are used to set values in `example:` struct tags
- All `RefOf()` calls have been replaced with `NewNullableFrom()`
- Tests have been updated that verify these changes.

### Checklist

- [ ] I have run this code, and it appears to resolve the stated issue.
- [ ] This PR does not reduce total test coverage
- [ ] This PR has no user interface changes or has already received approval from product management to change the interface.
- [ ] Does this change require a Terraform schema change?
  - If so what is the ticket or PR #
- [ ] Make a [changie](https://github.com/OpsLevel/opslevel-go/blob/main/CONTRIBUTING.md#changie-change-log-generation) entry that explains the customer facing outcome of this change

### CODE REVIEW

File: actions_test.go [hunk 1]
```diff
@@ -83,7 +83,7 @@ func TestUpdateWebhookAction2(t *testing.T) {
 	// Arrange
 	testRequest := autopilot.NewTestRequest(
 		`mutation WebhookActionUpdate($input:CustomActionsWebhookActionUpdateInput!){customActionsWebhookActionUpdate(input: $input){webhookAction{{ template "custom_actions_request" }},errors{message,path}}}`,
-		`{"input":{"id": "123456789","description":"","headers":"{\"Accept\":\"application/json\"}"}}`,
+		`{"input":{"id": "123456789","description":null,"headers":"{\"Accept\":\"application/json\"}"}}`,
```
Comment (rocktavious) [inline actions_test.go:86]: This seems like a material change - i'm not sure we want this.

File: gen.go [hunk 2]
```diff
@@ -110,6 +109,97 @@ var stringTypeSuffixes = []string{
 	"yaml",
 }
 
+var enumExamples = map[string]string{
+	"AlertSourceStatusTypeEnum":                       opslevel.AllAlertSourceStatusTypeEnum[0],
+	"ApiDocumentSourceEnum":                           opslevel.AllApiDocumentSourceEnum[0],
+	"AlertSourceTypeEnum":                             opslevel.AllAlertSourceTypeEnum[0],
```
Comment (rocktavious) [inline gen.go:115]: This feels like a chicken and egg problem since these are generated data - how does this work when enum.go doesn't exist?

Comment (rocktavious) [reply gen.go:115]: Reading further down - i think you can just replace this with data from the schema if you build it up during schema parsing first.

File: gen.go [hunk 3]
```diff
@@ -891,6 +1001,161 @@ func renameMutation(s string) string {
...
+	case "Float":
+		convertedType += "float64"
+	case "Int":
+		convertedType += "int"
+	case "ISO8601DateTime":
+		convertedType += "iso8601.Time"
+	case "String":
+		convertedType += "string"
+	default:
+		convertedType += graphqlType
```
Comment (rocktavious) [inline gen.go:1097]: should we log a warning when this happens so we can be alerted to it?

File: gen.go [hunk 4]
```diff
@@ -891,6 +1001,161 @@ func renameMutation(s string) string {
...
+			goType = "*JsonString"
+		case "PropertyDefinition":
+			goType = "PropertyDefinitionId"
+		}
+	} else if fieldType.Name == "ServiceRepository" && goType == "Repository" {
+		goType = "RepositoryId"
+	}
+
+	return goType
+}
```
Comment (rocktavious) [inline gen.go:1153]: This logic is pretty hard to follow.  Something to mark for refactoring or thinking if there is a better way.  I don't have a good handle on all the context needed to understand this yet but feels like a missing design pattern we could leverage that would significantly clean this up.  Maybe the Open/Closed principle or the Liskov Substitution principle.

File: input.go [hunk 5]
```diff
@@ -24,41 +24,108 @@ type AlertSourceServiceDeleteInput struct {
...
+	SubscriptionId        *Nullable[string]   `json:"subscriptionId,omitempty" yaml:"subscriptionId,omitempty" example:"example_value"`                // The subscription OpsLevel uses to access the Azure account. (Optional.)
+	TagsOverrideOwnership *Nullable[bool]     `json:"tagsOverrideOwnership,omitempty" yaml:"tagsOverrideOwnership,omitempty" example:"false"`          // Allow tags imported from Azure to override ownership set in OpsLevel directly. (Optional.)
+	TenantId              *Nullable[string]   `json:"tenantId,omitempty" yaml:"tenantId,omitempty" example:"example_value"`                            // The tenant OpsLevel uses to access the Azure account. (Optional.)
+}
+
+// CampaignCreateInput specifies the input fields used to create a campaign.
+type CampaignCreateInput struct {
+	CheckIdsToCopy *Nullable[[]ID]                  `json:"checkIdsToCopy,omitempty" yaml:"checkIdsToCopy,omitempty" example:"[]"`                  // The IDs of the existing rubric checks to be copied. (Optional.)
```
Comment (rocktavious) [inline input.go:60]: I think this might be an edgecase of not really truely being "optional". Also i'm not sure nullable of a slice makes sense.

File: input.go [hunk 6]
```diff
@@ -24,41 +24,108 @@ type AlertSourceServiceDeleteInput struct {
...
+	OwnershipTagKeys      *Nullable[[]string] `json:"ownershipTagKeys,omitempty" yaml:"ownershipTagKeys,omitempty" example:"['tag_key1', 'tag_key2']"` // An array of tag keys used to associate ownership from an integration. Max 5. (Optional.)
+	SubscriptionId        *Nullable[string]   `json:"subscriptionId,omitempty" yaml:"subscriptionId,omitempty" example:"example_value"`                // The subscription OpsLevel uses to access the Azure account. (Optional.)
+	TagsOverrideOwnership *Nullable[bool]     `json:"tagsOverrideOwnership,omitempty" yaml:"tagsOverrideOwnership,omitempty" example:"false"`          // Allow tags imported from Azure to override ownership set in OpsLevel directly. (Optional.)
+	TenantId              *Nullable[string]   `json:"tenantId,omitempty" yaml:"tenantId,omitempty" example:"example_value"`                            // The tenant OpsLevel uses to access the Azure account. (Optional.)
+}
+
+// CampaignCreateInput specifies the input fields used to create a campaign.
+type CampaignCreateInput struct {
```
Comment (rocktavious) [inline input.go:59]: We should probably put in an exclusion for everything "campaigns" related.  The API has never been vetted throughly for external use.  Its entirely possible that these structures don't work correctly.

File: input.go [hunk 7]
```diff
@@ -6,15 +6,15 @@ import "github.com/relvacode/iso8601"
 
 // AlertSourceExternalIdentifier specifies the input needed to find an alert source with external information.
 type AlertSourceExternalIdentifier struct {
-	Type       AlertSourceTypeEnum `json:"type" yaml:"type" example:"pagerduty"`                                   // The type of the alert. (Required.)
-	ExternalId string              `json:"externalId" yaml:"externalId" example:"Z2lkOi8vc2VydmljZS8xMjM0NTY3ODk"` // The external id of the alert. (Required.)
+	ExternalId string              `json:"externalId" yaml:"externalId" example:"example_value"` // The external id of the alert. (Required.)
```
Comment (rocktavious) [inline input.go:9]: why is the example value in the struct tag changing?

Comment (davidbloss) [reply input.go:9]: This is an edge case that needs special handling. The `Z2lk...` value is populated for fields of type `Id`. In the API schema this is string but I believe we expect it to have a value like an Id under the hood.

File: input.go (hunk 5 above)
Comment (davidbloss) [reply input.go:60]: This is definitely an odd case, but from a purely JSON perspective it makes sense. Unless some JSON schema forbids it, `null` fits pretty much anywhere. The mystery then is how it's handled on the app side

File: gen.go (hunk 4 above)
Comment (davidbloss) [reply gen.go:1153]: I'll work on this

File: gen.go (hunk 3 above)
Comment (davidbloss) [reply gen.go:1097]: I don't think a warning is needed here. The `default` case catches non-scalar API Schema defined types like `IdentifierInput`, `UserIdentifierInput`, `ToolCategory`, `AlertSourceExternalIdentifier`, etc

File: gen.go (hunk 2 above)
Comment (davidbloss) [reply gen.go:115]: I've got a simple fix for this 👍

File: input.go (hunk 6 above)
Comment (davidbloss) [reply input.go:59]: Campaign related objects no longer generated ✅

File: gen.go (hunk 4 above)
Comment (davidbloss) [reply gen.go]: This function is not used in this PR. Resolving this thread.

File: actions.go [hunk 8]
```diff
@@ -165,10 +165,12 @@ func (client *Client) CreateTriggerDefinition(input CustomActionsTriggerDefiniti
 		} `graphql:"customActionsTriggerDefinitionCreate(input: $input)"`
 	}
 	if input.AccessControl == nil {
-		input.AccessControl = RefOf(CustomActionsTriggerDefinitionAccessControlEnumEveryone)
+		everyoneEnum := CustomActionsTriggerDefinitionAccessControlEnumEveryone
+		input.AccessControl = &everyoneEnum
```
Comment (rocktavious) [inline actions.go:169]: Did you test if changing the enum to `var` instead of `const` makes is so we don't have to do this?

It should be a transparent change to the end user and make it so you can just do

```suggestion
		input.AccessControl = &CustomActionsTriggerDefinitionAccessControlEnumEveryone
```

Comment (davidbloss) [reply actions.go:169]: That does the trick. Should I open a separate PR to make this change to enum.go?

File: check_test.go [hunk 9]
```diff
@@ -346,31 +349,31 @@ func getCheckTestCases() map[string]TmpCheckTestCase {
 			fixture: BuildUpdateRequest("CodeIssue", map[string]any{
 				"constraint":     ol.CheckCodeIssueConstraintEnumContains,
 				"issueName":      "code-issue-updated",
-				"issueType":      nil, // API NOTE - not allowed when constraint is "contains"
+				"issueType":      &[]string{}, // API NOTE - not allowed when constraint is "contains"
```
Comment (rocktavious) [inline check_test.go:352]: shouldn't this be `(*[]string)(nil)` to be the same as before?  I guess i would expect `&[]string{}` to be an initialized by empty slice and thus generating `[]` ?

File: actions.go (hunk 8 above)
Comment (davidbloss) [reply actions.go]: Done ✅

File: check_test.go [hunk 10]
```diff
@@ -404,7 +407,7 @@ func getCheckTestCases() map[string]TmpCheckTestCase {
 			fixture: BuildUpdateRequest("HasRecentDeploy", map[string]any{"days": 5}),
 			body: func(c *ol.Client) (*ol.Check, error) {
 				input := ol.NewCheckUpdateInputTypeOf[ol.CheckHasRecentDeployUpdateInput](checkUpdateInput)
-				input.Days = ol.RefOf(5)
+				input.Days = ol.NewNullableFrom(5)
```
Comment (rocktavious) [inline check_test.go:408]: heres an `int` where i'm not sure `null` makes sense - can we test the API in Graphiql?

File: actions_test.go [hunk 11]
```diff
@@ -69,9 +71,10 @@ func TestUpdateWebhookAction(t *testing.T) {
 	client := BestTestClient(t, "custom_actions/update_action", testRequest)
 
 	// Act
+	putHttpMethod := ol.CustomActionsHttpMethodEnumPut
```
Comment (rocktavious) [inline actions_test.go:74]: missed

File: common.go [hunk 12]
```diff
@@ -22,11 +22,12 @@ type PayloadVariables map[string]interface{}
 
 // WithoutDeactivedUsers filters out deactivated users on ListUsers query
 func (pv *PayloadVariables) WithoutDeactivedUsers() *PayloadVariables {
+	enumEquals := BasicTypeEnumEquals
```
Comment (rocktavious) [inline common.go:25]: missed

File: actions_test.go [hunk 13]
```diff
@@ -186,12 +190,12 @@ func TestCreateTriggerDefinitionWithNullExtendedTeams(t *testing.T) {
 	trigger, err := client.CreateTriggerDefinition(ol.CustomActionsTriggerDefinitionCreateInput{
 		Name:                   "Deploy Rollback",
 		Description:            ol.RefOf("Disables the Deploy Freeze"),
-		AccessControl:          ol.RefOf(ol.CustomActionsTriggerDefinitionAccessControlEnumEveryone),
+		AccessControl:          &ol.CustomActionsTriggerDefinitionAccessControlEnumEveryone,
 		ManualInputsDefinition: ol.RefOf(""),
 		ResponseTemplate:       ol.RefOf(""),
-		ActionId:               newID,
+		ActionId:               ol.RefOf(*newID),
```
Comment (rocktavious) [inline actions_test.go:196]: this is uncessary?

File: actions_test.go [hunk 14]
```diff
@@ -130,13 +134,13 @@ func TestCreateTriggerDefinition(t *testing.T) {
 	// Act
 	trigger, err := client.CreateTriggerDefinition(ol.CustomActionsTriggerDefinitionCreateInput{
 		Name:                   "Deploy Rollback",
-		AccessControl:          ol.RefOf(ol.CustomActionsTriggerDefinitionAccessControlEnumEveryone),
+		AccessControl:          &ol.CustomActionsTriggerDefinitionAccessControlEnumEveryone,
 		Description:            ol.RefOf("Disables the Deploy Freeze"),
 		ResponseTemplate:       ol.RefOf(""),
 		ManualInputsDefinition: ol.RefOf(""),
-		ActionId:               newID,
+		ActionId:               ol.RefOf(*newID),
```
Comment (rocktavious) [inline actions_test.go:141]: unecessary?

File: actions_test.go [hunk 15]
```diff
@@ -155,14 +159,14 @@ func TestCreateTriggerDefinitionWithGlobalEntityType(t *testing.T) {
 	// Act
 	trigger, err := client.CreateTriggerDefinition(ol.CustomActionsTriggerDefinitionCreateInput{
 		Name:                   "Deploy Rollback",
-		AccessControl:          ol.RefOf(ol.CustomActionsTriggerDefinitionAccessControlEnumEveryone),
+		AccessControl:          &ol.CustomActionsTriggerDefinitionAccessControlEnumEveryone,
 		Description:            ol.RefOf("Disables the Deploy Freeze"),
-		ActionId:               newID,
+		ActionId:               ol.RefOf(*newID),
```
Comment (rocktavious) [inline actions_test.go:164]: unecessary?

File: check_test.go [hunk 16]
```diff
@@ -32,8 +32,9 @@ var (
 		Value: ol.RefOf("Requests"),
 	}
 
+	equalsEnum           = ol.PredicateTypeEnumEquals
```
Comment (rocktavious) [inline check_test.go:35]: missed?

File: input.go [hunk 17]
```diff
@@ -68,586 +68,609 @@ type CategoryDeleteInput struct {
...
+	MaxAllowed     *Nullable[int]                `json:"maxAllowed,omitempty" yaml:"maxAllowed,omitempty" example:"3"`                               // The threshold count of code issues beyond which the check starts failing. (Optional.)
+	Name           *Nullable[string]             `json:"name,omitempty" yaml:"name,omitempty" example:"example_value"`                               // The display name of the check. (Optional.)
+	Notes          *Nullable[string]             `json:"notes,omitempty" yaml:"notes,omitempty" example:"example_value"`                             // Additional information about the check. (Optional.)
+	OwnerId        *Nullable[ID]                 `json:"ownerId,omitempty" yaml:"ownerId,omitempty" example:"Z2lkOi8vc2VydmljZS8xMjM0NTY3ODk"`       // The id of the owner of the check. (Optional.)
+	ResolutionTime *CodeIssueResolutionTimeInput `json:"resolutionTime,omitempty" yaml:"resolutionTime,omitempty"`                                   // The resolution time recommended by the reporting source of the code issue. (Optional.)
```
Comment (rocktavious) [inline input.go:138]: This used to be nullable - but now isn't?

File: input.go [hunk 18]
```diff
@@ -68,586 +68,609 @@ type CategoryDeleteInput struct {
...
+	LevelId               *Nullable[ID]                    `json:"levelId,omitempty" yaml:"levelId,omitempty" example:"Z2lkOi8vc2VydmljZS8xMjM0NTY3ODk"`       // The id of the level the check belongs to. (Optional.)
+	Name                  *Nullable[string]                `json:"name,omitempty" yaml:"name,omitempty" example:"example_value"`                               // The display name of the check. (Optional.)
+	Notes                 *Nullable[string]                `json:"notes,omitempty" yaml:"notes,omitempty" example:"example_value"`                             // Additional information about the check. (Optional.)
+	OwnerId               *Nullable[ID]                    `json:"ownerId,omitempty" yaml:"ownerId,omitempty" example:"Z2lkOi8vc2VydmljZS8xMjM0NTY3ODk"`       // The id of the owner of the check. (Optional.)
 	UpdateFrequency       *ManualCheckFrequencyUpdateInput `json:"updateFrequency,omitempty" yaml:"updateFrequency,omitempty"`                                 // Defines the minimum frequency of the updates. (Optional.)
```
Comment (rocktavious) [inline input.go:296]: This field appears to be optional but isn't wrapped in Nullable?

File: input.go [hunk 19]
```diff
@@ -68,586 +68,609 @@ type CategoryDeleteInput struct {
...
-	Notes                 *string                    `json:"notes,omitempty" yaml:"notes,omitempty" example:"example_notes"`                         // Additional information about the check. (Optional.)
+	Name                  string                     `json:"name" yaml:"name" example:"example_value"`                                               // The display name of the check. (Required.)
+	Notes                 *Nullable[string]          `json:"notes,omitempty" yaml:"notes,omitempty" example:"example_value"`                         // Additional information about the check. (Optional.)
+	OwnerId               *Nullable[ID]              `json:"ownerId,omitempty" yaml:"ownerId,omitempty" example:"Z2lkOi8vc2VydmljZS8xMjM0NTY3ODk"`   // The id of the team that owns the check. (Optional.)
 	UpdateFrequency       *ManualCheckFrequencyInput `json:"updateFrequency,omitempty" yaml:"updateFrequency,omitempty"`                             // Defines the minimum frequency of the updates. (Optional.)
```
Comment (rocktavious) [inline input.go:281]: This field appears to be optional but isn't wrapped in Nullable?

File: check_test.go [hunk 20]
```diff
@@ -1161,6 +1170,9 @@ func TestJsonUnmarshalCreateCheckManual(t *testing.T) {
 	// Act
 	buf1, err1 := ol.UnmarshalCheckCreateInput(ol.CheckTypeManual, []byte(data))
 	// Assert
+	stuff := buf1.(*ol.CheckManualCreateInput)
+	stuff2 := &output
+	fmt.Println(stuff2, stuff)
```
Comment (rocktavious) [inline check_test.go:1175]:
```suggestion
```

File: scalar_test.go [hunk 21]
```diff
@@ -224,7 +224,7 @@ func TestNewIdentifierArray(t *testing.T) {
 }
 
 func TestNewNullString(t *testing.T) {
-	buf, err := json.Marshal(ol.NewNull[string]())
```
Comment (rocktavious) [inline scalar_test.go:227]: I think the changes in this file can be undone

File: tags.go [hunk 22]
```diff
@@ -222,9 +222,10 @@ func (client *Client) ReconcileTags(resourceType TaggableResourceInterface, tags
 
 	toCreate, toDelete := reconcileTags(tagConnection.Nodes, tagsDesired)
 	for _, tag := range toCreate {
+		taggableResourceType := resourceType.ResourceType()
```
Comment (rocktavious) [inline tags.go:225]: Its unclear to me why this is changing.

File: user_test.go [hunk 23]
```diff
@@ -38,9 +38,10 @@ func TestInviteUserSkipSendInvite(t *testing.T) {
 	)
 
 	client := BestTestClient(t, "user/invite_skip_send_invite", testRequest)
+	teamMemberRole := ol.UserRoleTeamMember
```
Comment (rocktavious) [inline user_test.go:41]: missed?

**Review by rocktavious — APPROVED** [review]
🚀
